sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,351 @@
+	case PluginTypeCredential:
+		return "auth"
+	case PluginTypeDatabase:
+		return "database"
+	case PluginTypeSecrets:
+		return "secret"
+	default:
+		return "unsupported"
+	}
+}
+
+// EnvironmentOptions are a collection of options each step driver should
+// support
+type EnvironmentOptions struct {
+	// MountPathPrefix is an optional prefix to use when mounting the plugin. If
+	// omitted the mount path will default to the PluginName with a random suffix.
+	MountPathPrefix string
+
+	// Name is used to register the plugin. This can be arbitray but should be a
+	// reasonable value. For an example, if the plugin in test is a secret backend
+	// that generates UUIDs with the name "vault-plugin-secrets-uuid", then "uuid"
+	// or "test-uuid" would be reasonable. The name is used for lookups in the
+	// catalog. See "name" in the "Register Plugin" endpoint docs:
+	// - https://www.vaultproject.io/api-docs/system/plugins-catalog#register-plugin
+	Name string
+
+	// PluginType is the optional type of plugin. See PluginType const defined
+	// above
+	PluginType PluginType
+
+	// PluginName represents the name of the plugin that gets compiled. In the
+	// standard plugin project file layout, it represents the folder under the
+	// cmd/ folder. In the below example UUID project, the PluginName would be
+	// "uuid":
+	//
+	// vault-plugin-secrets-uuid/
+	// - backend.go
+	// - cmd/
+	// ----uuid/
+	// ------main.go
+	// - path_generate.go
+	//
+	PluginName string
+}
+
+// Step represents a single step of a test Case
+type Step struct {
+	// Operation defines what action is being taken in this step; write, read,
+	// delete, et. al.
+	Operation StepOperation
+
+	// Path is the localized request path. The mount prefix, namespace, and
+	// optionally "auth" will be automatically added.
+	Path string
+
+	// Arguments to pass in the request. These arguments represent payloads sent
+	// to the API.
+	Data map[string]interface{}
+
+	// Check is a function that is called after this step is executed in order to
+	// test that the step executed successfully. If this is not set, then the next
+	// step will be called
+	Check StepCheckFunc
+
+	// ExpectError indicates if this step is expected to return an error. If the
+	// step operation returns an error and ExpectError is not set, the test will
+	// fail immediately and the StepCheckFunc will not be ran. If ExpectError is
+	// true, the StepCheckFunc will be called (if any) and include the error from
+	// the response. It is the responsibility of the StepCheckFunc to validate the
+	// error is appropriate or not, if expected.
+	ExpectError bool
+
+	// Unauthenticated will make the request unauthenticated.
+	Unauthenticated bool
+}
+
+// StepCheckFunc is the callback used for Check in Steps.
+type StepCheckFunc func(*api.Secret, error) error
+
+// Case represents a scenario we want to test which involves a series of
+// steps to be followed sequentially, evaluating the results after each step.
+type Case struct {
```

Comment:
I originally named it `Case` to differentiate from the legacy test `TestCase`. In hindsight I think I should move the legacy files to their own package at the same level as `stepwise` and split the two, because right now the legacy testing is a part of the `stepwise` package. I'm going to defer that change and make a note internally to do so after this PR is merged, otherwise it will introduce a new set of changes here.